### PR TITLE
SWATCH-4878 Add EventNormalizer filtering for non-PAYG products

### PR DESCRIPTION
**Pull request by mstead**
Jira issue: SWATCH-4878

## Description
Filter any non-paygo products and metrics from incoming events to prevent traditional products from being tallied durring the hourly tally process.

- Add filterNonPaygoTags() to remove TRADITIONAL product tags from events during normalization, ensuring only PAYG products are processed during the hourly tally.

Added swatch-tally component tests for PAYG filtering

- Add TallyFilterNonPaygoProductsTest with 5 comprehensive test cases: TC001: Mixed PAYG/TRADITIONAL tags - verifies filtering and measurements TC002: TRADITIONAL-only events excluded from hourly processing TC003: PAYG-only events processed normally (baseline) TC004: Multiple events with mixed tags filtered correctly TC005: Conflict resolution with mixed tags (same timestamp)
- Add test documentation to TEST_PLAN.md

- Add unit test coverage for EventNormalizer

- Add unit test coverage around the EventNormalizer)

### Reproducer

1. Remove the clutter and start with a fresh DB
	```
	dropdb -h localhost -U rhsm-subscriptions rhsm-subscriptions && \
	createdb -h localhost -U rhsm-subscriptions rhsm-subscriptions
	```
2. Check out the `main` branch and deploy `make swatch-tally`
3. Send the following event message which contains a mix of paygo and traditional product data.
```
http :9080/topics/platform.rhsm-subscriptions.service-instance-ingress \
 Content-Type:application/vnd.kafka.json.v2+json \
 Accept:application/vnd.kafka.v2+json --raw '{
  "records": [
	{
		"key": "11111111",
		"value": {
			"service_type": "RHEL System",
            "timestamp": "2026-06-03T14:00:00Z",
            "expiration": "2026-06-03T15:00:00Z",
			"display_name": "test-payg-filter-1",
			"instance_id": "999999999",
			"org_id": "11111111",
			"product_ids": ["479"],
			"product_tag": ["rhel-for-x86-els-payg-addon", "rhel-for-x86-els-unconverted"],
			"measurements": [
				{
					"metric_id": "vCPUs",
					"value": 4.0
				},
				{
					"metric_id": "Sockets",
					"value": 2.0
				}
			],
			"billing_provider": "aws",
			"billing_account_id": "123456789012",
			"sla": "Premium",
			"usage": "Production",
			"event_source": "prometheus",
			"event_type": "snapshot"
		}
	}
  ]
}'
```

Notice that 4 events were created! An event was created for each traditional and paygo product mapped to each measurement which is incorrect. There should just be 1 for paygo.
```
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions \
-c "select instance_id, timestamp, data->'product_tag', data->'measurements'->0 from events where org_id='11111111' order by timestamp desc;"

```
```
 instance_id |       timestamp        |             ?column?             |                ?column?                
-------------+------------------------+----------------------------------+----------------------------------------
 999999999   | 2026-06-03 14:00:00+00 | ["rhel-for-x86-els-unconverted"] | {"value": 2.0, "metric_id": "Sockets"}
 999999999   | 2026-06-03 14:00:00+00 | ["rhel-for-x86-els-unconverted"] | {"value": 4.0, "metric_id": "vCPUs"}
 999999999   | 2026-06-03 14:00:00+00 | ["rhel-for-x86-els-payg-addon"]  | {"value": 2.0, "metric_id": "Sockets"}
 999999999   | 2026-06-03 14:00:00+00 | ["rhel-for-x86-els-payg-addon"]  | {"value": 4.0, "metric_id": "vCPUs"}
(4 rows)
```

3. Run the hourly tally
```
http POST ":8000/api/rhsm-subscriptions/v1/internal/tally/hourly?org=11111111" x-rh-swatch-psk:placeholder
```
4. Notice that monthly totals were set for SOCKETS and VCPUS.
```
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions -c "select * from instance_measurements order by last_modified desc"
```

```
               host_id                | metric_id | value |             instance_id              |         last_modified         
--------------------------------------+-----------+-------+--------------------------------------+-------------------------------
 e09782c6-9633-4f25-8c82-e1125c68d562 | SOCKETS   |     2 | e09782c6-9633-4f25-8c82-e1125c68d562 | 2026-06-22 17:42:27.646886+00
 e09782c6-9633-4f25-8c82-e1125c68d562 | VCPUS     |     4 | e09782c6-9633-4f25-8c82-e1125c68d562 | 2026-06-22 17:42:27.646886+00
(2 rows)
```

Notice that snapshots/measurements were created for both products.
```
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions -c "select snapshot_date, granularity, product_id, sla, usage, billing_provider, billing_account_id, m.metric_id, m.value from tally_snapshots s inner join tally_measurements m on m.snapshot_id=s.id where s.org_id='11111111' and s.is_primary=true order by metric_id, granularity"
```

```
     snapshot_date      | granularity |          product_id          |   sla   |   usage    | billing_provider | billing_account_id | metric_id | value 
------------------------+-------------+------------------------------+---------+------------+------------------+--------------------+-----------+-------
 2026-06-03 00:00:00+00 | DAILY       | rhel-for-x86-els-unconverted | Premium | Production | _ANY             | _ANY               | SOCKETS   |     2
 2026-06-03 00:00:00+00 | DAILY       | rhel-for-x86-els-unconverted | Premium | Production | _ANY             | _ANY               | SOCKETS   |     2
 2026-06-03 14:00:00+00 | HOURLY      | rhel-for-x86-els-unconverted | Premium | Production | _ANY             | _ANY               | SOCKETS   |     2
 2026-06-03 14:00:00+00 | HOURLY      | rhel-for-x86-els-unconverted | Premium | Production | _ANY             | _ANY               | SOCKETS   |     2
 2026-06-03 00:00:00+00 | DAILY       | rhel-for-x86-els-payg-addon  | Premium | Production | aws              | 123456789012       | VCPUS     |     4
 2026-06-03 00:00:00+00 | DAILY       | rhel-for-x86-els-payg-addon  | Premium | Production | aws              | 123456789012       | VCPUS     |     4
 2026-06-03 14:00:00+00 | HOURLY      | rhel-for-x86-els-payg-addon  | Premium | Production | aws              | 123456789012       | VCPUS     |     4
 2026-06-03 14:00:00+00 | HOURLY      | rhel-for-x86-els-payg-addon  | Premium | Production | aws              | 123456789012       | VCPUS     |     4
```

### Verification
1. Clean your database and start fresh so it is easier to verify.
```
dropdb -h localhost -U rhsm-subscriptions rhsm-subscriptions && \
createdb -h localhost -U rhsm-subscriptions rhsm-subscriptions
```
2. Check out the PR's branch and deploy again.
3. Send the following event (mix)
```
http :9080/topics/platform.rhsm-subscriptions.service-instance-ingress \
 Content-Type:application/vnd.kafka.json.v2+json \
 Accept:application/vnd.kafka.v2+json --raw '{
  "records": [
	{
		"key": "11111111",
		"value": {
			"service_type": "RHEL System",
            "timestamp": "2026-06-03T14:00:00Z",
            "expiration": "2026-06-03T15:00:00Z",
			"display_name": "test-payg-filter-1",
			"instance_id": "999999999",
			"org_id": "11111111",
			"product_ids": ["479"],
			"product_tag": ["rhel-for-x86-els-payg-addon", "rhel-for-x86-els-unconverted"],
			"measurements": [
				{
					"metric_id": "vCPUs",
					"value": 4.0
				},
				{
					"metric_id": "Sockets",
					"value": 2.0
				}
			],
			"billing_provider": "aws",
			"billing_account_id": "123456789012",
			"sla": "Premium",
			"usage": "Production",
			"event_source": "prometheus",
			"event_type": "snapshot"
		}
	}
  ]
}'
```

Verify that now there's only 1 event created and that it is for the paygo product with vCPUs measurement.
```
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions \
-c "select instance_id, timestamp, data->'product_tag', data->'measurements'->0 from events where org_id='11111111' order by timestamp desc;"

```
```
 instance_id |       timestamp        |            ?column?             |               ?column?               
-------------+------------------------+---------------------------------+--------------------------------------
 999999999   | 2026-06-03 14:00:00+00 | ["rhel-for-x86-els-payg-addon"] | {"value": 4.0, "metric_id": "vCPUs"}
(1 row)

```

4. Run the hourly tally.
```
http POST ":8000/api/rhsm-subscriptions/v1/internal/tally/hourly?org=11111111" x-rh-swatch-psk:placeholder
```

Notice that monthly totals were set for VCPUS only.
```
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions -c "select * from instance_measurements order by last_modified desc"
```

```
               host_id                | metric_id | value |             instance_id              |         last_modified         
--------------------------------------+-----------+-------+--------------------------------------+-------------------------------
 5a48a465-8d75-4499-82ef-79113a996db7 | VCPUS     |     4 | 5a48a465-8d75-4499-82ef-79113a996db7 | 2026-06-22 18:04:41.426973+00
(1 row)
```

Also notice that snapshots and measurements are only created for the paygo product and measurement.

```
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions -c "select snapshot_date, granularity, product_id, sla, usage, billing_provider, billing_account_id, m.metric_id, m.value from tally_snapshots s inner join tally_measurements m on m.snapshot_id=s.id where s.org_id='11111111' and s.is_primary=true order by metric_id, granularity"
```

```
     snapshot_date      | granularity |         product_id          |   sla   |   usage    | billing_provider | billing_account_id | metric_id | value 
------------------------+-------------+-----------------------------+---------+------------+------------------+--------------------+-----------+-------
 2026-06-03 00:00:00+00 | DAILY       | rhel-for-x86-els-payg-addon | Premium | Production | aws              | 123456789012       | VCPUS     |     4
 2026-06-03 00:00:00+00 | DAILY       | rhel-for-x86-els-payg-addon | Premium | Production | aws              | 123456789012       | VCPUS     |     4
 2026-06-03 14:00:00+00 | HOURLY      | rhel-for-x86-els-payg-addon | Premium | Production | aws              | 123456789012       | VCPUS     |     4
 2026-06-03 14:00:00+00 | HOURLY      | rhel-for-x86-els-payg-addon | Premium | Production | aws              | 123456789012       | VCPUS     |     4
(4 rows)
```

#### Further Verification
We will stop verification at the Event level since we have the base case covered. We just need to ensure that the correct event records are produced in the DB. The has been no change to the tally process, just event ingestion.

Use the same event query to verify the events as they are processed:
```
psql -h localhost -U rhsm-subscriptions rhsm-subscriptions \
-c "select instance_id, timestamp, data->'product_tag', data->'measurements'->0 from events where org_id='11111111' order by timestamp desc;"

```

**Conflict Resolution still functions**

  ```
  http :9080/topics/platform.rhsm-subscriptions.service-instance-ingress \
 Content-Type:application/vnd.kafka.json.v2+json \
 Accept:application/vnd.kafka.v2+json --raw '{
  "records": [
	{
		"key": "11111111",
		"value": {
			"service_type": "RHEL System",
            "timestamp": "2026-06-03T14:00:00Z",
            "expiration": "2026-06-03T15:00:00Z",
			"display_name": "test-payg-filter-1",
			"instance_id": "999999999",
			"org_id": "11111111",
			"product_ids": ["479"],
			"product_tag": ["rhel-for-x86-els-payg-addon", "rhel-for-x86-els-unconverted"],
			"measurements": [
				{
					"metric_id": "vCPUs",
					"value": 10.0
				},
				{
					"metric_id": "Sockets",
					"value": 2.0
				}
			],
			"billing_provider": "aws",
			"billing_account_id": "123456789012",
			"sla": "Premium",
			"usage": "Production",
			"event_source": "prometheus",
			"event_type": "snapshot"
		}
	}
  ]
}'
  ```

Hourly tally will produce 2 new events:
* 1 deduction event: -4 vCPUs
* 1 amendment showing a value of 10.0 vCPUs
* No events for Traditional product produced.

**Traditional Product Events are skipped entirely**

```
  http :9080/topics/platform.rhsm-subscriptions.service-instance-ingress \
 Content-Type:application/vnd.kafka.json.v2+json \
 Accept:application/vnd.kafka.v2+json --raw '{
  "records": [
	{
		"key": "11111111",
		"value": {
			"service_type": "RHEL System",
            "timestamp": "2026-06-03T15:00:00Z",
            "expiration": "2026-06-03T16:00:00Z",
			"display_name": "test-payg-filter-1",
			"instance_id": "999999999",
			"org_id": "11111111",
			"product_ids": ["479"],
			"product_tag": ["rhel-for-x86-els-unconverted"],
			"measurements": [
				{
					"metric_id": "Sockets",
					"value": 2.0
				}
			],
			"billing_provider": "aws",
			"billing_account_id": "123456789012",
			"sla": "Premium",
			"usage": "Production",
			"event_source": "prometheus",
			"event_type": "snapshot"
		}
	}
  ]
}'
```

No event produced since there are no applicable products/measurements to apply since it is traditional.

**Normal PayGo Event Processing Continues To Work**

```
  http :9080/topics/platform.rhsm-subscriptions.service-instance-ingress \
 Content-Type:application/vnd.kafka.json.v2+json \
 Accept:application/vnd.kafka.v2+json --raw '{
  "records": [
	{
		"key": "11111111",
		"value": {
			"service_type": "RHEL System",
            "timestamp": "2026-06-03T18:00:00Z",
            "expiration": "2026-06-03T19:00:00Z",
			"display_name": "test-payg-filter-1",
			"instance_id": "999999999",
			"org_id": "11111111",
			"product_ids": ["479"],
			"product_tag": ["rhel-for-x86-els-payg-addon"],
			"measurements": [
				{
					"metric_id": "vCPUs",
					"value": 16.0
				}
			],
			"billing_provider": "aws",
			"billing_account_id": "123456789012",
			"sla": "Premium",
			"usage": "Production",
			"event_source": "prometheus",
			"event_type": "snapshot"
		}
	}
  ]
}'
```

This will produce a new event for the specified timestamp (basic normal functionality).

The final event records should look as follows:
```
 instance_id |       timestamp        |            ?column?             |               ?column?                
-------------+------------------------+---------------------------------+---------------------------------------
 999999999   | 2026-06-03 18:00:00+00 | ["rhel-for-x86-els-payg-addon"] | {"value": 16.0, "metric_id": "vCPUs"}
 999999999   | 2026-06-03 14:00:00+00 | ["rhel-for-x86-els-payg-addon"] | {"value": 10.0, "metric_id": "vCPUs"}
 999999999   | 2026-06-03 14:00:00+00 | ["rhel-for-x86-els-payg-addon"] | {"value": -4.0, "metric_id": "vCPUs"}
 999999999   | 2026-06-03 14:00:00+00 | ["rhel-for-x86-els-payg-addon"] | {"value": 4.0, "metric_id": "vCPUs"}
(4 rows)
```

## Summary of Changes

* **Behavior Changes / New Features**
  * Improved event normalization and flattening to enforce PAYGO eligibility, clearing non-PAYGO tags and removing unsupported tag/measurement combinations for downstream processing.
  * Added PAYGO-focused handling for service instance persistence, ensuring only PAYGO-tagged values and supported measurements are kept.
  * Continued service type conversion for “Ansible Infrastructure Hour” to “Ansible Managed Node”.

* **Tests**
  * Added/expanded unit tests for PAYGO tag and measurement filtering, service type conversion, and flattened event expansion, including mixed/edge cases.
  * Added component tests and an hourly tally test-plan section covering PAYGO vs non-PAYGO filtering, conflict resolution, and role-based tag derivation (TC007).

* **Chores / Test Maintenance**
  * Refreshed shared tally test helpers and updated tally/conflict tests to use common await/query utilities.
  * Updated conflict-resolver test constants to domain-specific values.

### CODE REVIEW

File: swatch-tally/TEST_PLAN.md [hunk 1]
```diff
@@ -118,6 +118,105 @@ Test cases should be testable locally and in deployed environments.
...
+    - After first tally: PAYG product has tally sum of 4.0
+    - After second tally: PAYG product has tally sum of 8.0 (conflict resolved to higher value)
+    - PAYG instance measurements updated to 8.0
+    - TRADITIONAL product has 0 instances in both tally runs
+- **Expected Result**:
+    - Conflict resolution works correctly with filtered PAYG tags
+    - PAYG product measurements are updated to the higher value
+    - TRADITIONAL tags remain filtered out during conflict resolution
+    - Instance data reflects the resolved measurement value
+
```
[the remark this answers is not in the document]
Comment (mstead) [reply swatch-tally/TEST_PLAN.md:219]: Done

File: src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java [hunk 2]
```diff
@@ -71,9 +79,73 @@ public Event normalizeEvent(Event event) {
...
+
+    Set<String> paygoTags =
+        productTags.stream().filter(this::isPaygoEligibleTag).collect(Collectors.toSet());
+
+    if (paygoTags.isEmpty()) {
+      log.debug("No paygo-eligible tags found in {}. Clearing product_tag.", productTags);
+      event.setProductTag(null);
+    } else {
+      if (paygoTags.size() < productTags.size()) {
+        log.warn("Filtered non-paygo tags from {}. Keeping: {}", productTags, paygoTags);
```
Comment (awood) [inline src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java]: Should this be at `warn` level?  Are mixed tags anything we can actually do anything about?

File: src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java [hunk 3]
```diff
@@ -71,9 +79,73 @@ public Event normalizeEvent(Event event) {
...
+    List<Measurement> validMeasurements =
+        measurements.stream()
+            .filter(
+                m -> supportedMetrics.contains(MetricIdUtils.toUpperCaseFormatted(m.getMetricId())))
+            .toList();
+
+    if (validMeasurements.size() < measurements.size()) {
+      List<Measurement> filteredOut =
+          measurements.stream().filter(m -> !supportedMetrics.contains(m.getMetricId())).toList();
+      log.warn(
```
Comment (awood) [inline src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java]: Same question about `warn` level.

File: src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java [hunk 4]
```diff
@@ -71,9 +79,73 @@ public Event normalizeEvent(Event event) {
...
+    } else {
+      if (paygoTags.size() < productTags.size()) {
+        log.warn("Filtered non-paygo tags from {}. Keeping: {}", productTags, paygoTags);
+        event.setProductTag(paygoTags);
+      }
+      filterUnsupportedMeasurements(event, paygoTags);
+    }
+  }
+
+  private void filterUnsupportedMeasurements(Event event, Set<String> paygoTags) {
```
Comment (awood) [inline src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java]: I'd like to make this not a `void` method.  Having a method without side-effects will make things easier if we ever decide to unit test this.

```patch
diff --git i/src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java w/src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java
index ad64e65b3..092d72bd0 100644
--- i/src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java
+++ w/src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java
@@ -104,14 +104,14 @@ public class EventNormalizer {
         log.warn("Filtered non-paygo tags from {}. Keeping: {}", productTags, paygoTags);
         event.setProductTag(paygoTags);
       }
-      filterUnsupportedMeasurements(event, paygoTags);
+      event.setMeasurements(filterUnsupportedMeasurements(event.getMeasurements(), paygoTags));
     }
   }
 
-  private void filterUnsupportedMeasurements(Event event, Set<String> paygoTags) {
-    List<Measurement> measurements = event.getMeasurements();
+  private List<Measurement> filterUnsupportedMeasurements(
+      List<Measurement> measurements, Set<String> paygoTags) {
     if (measurements == null || measurements.isEmpty()) {
-      return;
+      return measurements;
     }
 
     // Get all supported metrics for the remaining paygo tags
@@ -135,7 +135,9 @@ public class EventNormalizer {
           paygoTags,
           filteredOut.stream().map(Measurement::getMetricId).collect(Collectors.toList()),
           validMeasurements.stream().map(Measurement::getMetricId).collect(Collectors.toList()));
-      event.setMeasurements(validMeasurements);
+      return validMeasurements;
+    } else {
+      return measurements;
     }
   }
```

And on that note, how about a few unit tests here?  Make these methods `protected` so we cover the code during ordinary test runs.

File: src/test/java/org/candlepin/subscriptions/event/EventNormalizerTest.java [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+    event.setMeasurements(List.of(createMeasurement(VCPUS, 4.0), createMeasurement(SOCKETS, 2.0)));
+
+    // When: Normalizing the event
+    Event result = normalizer.normalizeEvent(event);
+
+    // Then: Only PAYG tags remain
+    assertNotNull(result.getProductTag(), "Product tags should not be null");
+    assertEquals(1, result.getProductTag().size(), "Should have only PAYG tag remaining");
+    assertTrue(
+        result.getProductTag().contains("rhel-for-x86-els-payg-addon"), "Should keep PAYG tag");
```
Comment (awood) [inline src/test/java/org/candlepin/subscriptions/event/EventNormalizerTest.java:68]: Can we just reduce these assertions to `result.getProductTag().equals(Set.of("rhel-for-x86-els-payg-addon"))`?

File: src/test/java/org/candlepin/subscriptions/event/EventNormalizerTest.java [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+    Event result = normalizer.normalizeEvent(event);
+
+    // Then: Tags and measurements unchanged
+    assertNotNull(result.getProductTag());
+    assertEquals(1, result.getProductTag().size());
+    assertTrue(result.getProductTag().contains("rhel-for-x86-els-payg-addon"));
+    assertNotNull(result.getMeasurements());
+    assertEquals(1, result.getMeasurements().size());
+    assertEquals(VCPUS.toUpperCaseFormatted(), result.getMeasurements().getFirst().getMetricId());
+    assertEquals(4.0, result.getMeasurements().getFirst().getValue());
```
Comment (awood) [inline src/test/java/org/candlepin/subscriptions/event/EventNormalizerTest.java:110]: Can we simplify all these assertions?  Just store the productTag set and measurement list as "expectedTags" and "expectedMeasurements" and then assertEquals(expectedTags, event.getProductTags), etc.  The `Measurement` class has a logical looking `equals` method generated (i.e. not just object identity)

**Review by awood — CHANGES_REQUESTED** [review]
I'd like to see the assertions in the EventNormalizerTest cleaned up a bit.  There's a lot of redundancy there: assert X is not null, assert X has size of 1, assert X's sole element is equal to "Y" instead of just assert X.getThing().equals(Y).  Likewise, we can use Measurement's equals implementation to reduce a lot of redundancy.

File: src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java (hunk 2 above)
Comment (mstead) [reply src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java]: Has been addressed

File: src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java (hunk 3 above)
Comment (mstead) [reply src/main/java/org/candlepin/subscriptions/event/EventNormalizer.java]: Has been addressed.

**Review by awood — DISMISSED** [review]
I like the redesign in 09c1cf0e5733af8cd626eb3d770d65e7427c9271!

**Review by liwalker-rh — APPROVED** [review]
Component tests look good. Also tried manually testing many scenarios, including some others involving role and everything looks as expected.
